### PR TITLE
XT-2860 Add ability to download original filing documents to ixbrl-viewer

### PR DESCRIPTION
**Pull request by sagesmith-wf**
#### Reason for change
- Other xbrl viewers offer the ability to download the original filing documents used to generate the viewer (see edgar renderer), so we will offer the ability as well

#### Description of change
- Addresses https://github.com/Workiva/ixbrl-viewer/issues/410 and https://github.com/Workiva/ixbrl-viewer/issues/409
- Adds the ability to save the viewer as a zip from both the gui and commandline versions of arelle
- When the xbrl model loaded into arelle is from a zip, the zip is now copied inline with the output of the viewer and users are given the ability to download the original unedited filing documents

## Steps to Test
Use the following  docs when testing:
[filing_documents.zip](https://github.com/Workiva/ixbrl-viewer/files/11378077/filing_documents.15.zip)

### Test the arelle command line
1. Pull this branch
2. Build the viewer `npm install && nmp run prod`
3. Do all of the following from your local arelle

#### Test with zip entrypoint commandline
For all of the following you will expect:
1. the filing documents to be outputted adjacent to the viewer
2. Have the download filing documents button available under the gear icon
3. See an entry for `filingDocuments` In the ixbrl-viewer.json
4. When the download filing documents button is hit, the saved zip file should have the same name as the original file as well as the same contents when unzipped.

##### Testing Steps
1. Download the zip
2. Test load from a zip and output to a directory
```bash
mkdir viewerDirectoryTest
python arelleCmdLine.py --plugins=<PATH_TO_VIEWER>/iXBRLViewerPlugin -f <PATH_TO_DOWNLOADS>/filing_documents.zip  --save-viewer <PATH_TO_OUTPUT>/viewerDirectoryTest --viewer-url <PATH_TO_VIEWER>/iXBRLViewerPlugin/viewer/dist/ixbrlviewer.js
echo ""
echo "Viewer directory test"
echo "$(ls viewerDirectoryTest)"
```
In the created `viewerDirectoryTest` you should see `ixbrlviewer.htm` and `filing-documents.zip`. 

3. Test load from a zip and output to a file
```bash
mkdir viewerFileTest
python arelleCmdLine.py --plugins=<PATH_TO_VIEWER>/iXBRLViewerPlugin -f <PATH_TO_DOWNLOADS>/filing_documents.zip  --save-viewer <PATH_TO_OUTPUT>/viewerFileTest/ixbrlviewer.html --viewer-url <PATH_TO_VIEWER>/iXBRLViewerPlugin/viewer/dist/ixbrlviewer.js
echo ""
echo "Viewer file test"
echo "$(ls viewerFileTest)"
```
In the created `viewerFileTest` you should see `ixbrlviewer.htm` and `filing-documents.zip`. 

4. Test load from a zip and output to a directory as a zip
```bash
mkdir viewerDirectoryZipTest
python arelleCmdLine.py --plugins=<PATH_TO_VIEWER>/iXBRLViewerPlugin -f <PATH_TO_DOWNLOADS>/filing_documents.zip  --save-viewer <PATH_TO_OUTPUT>/viewerDirectoryZipTest --viewer-url <PATH_TO_VIEWER>/iXBRLViewerPlugin/viewer/dist/ixbrlviewer.js --zip-viewer-output
echo ""
echo "Viewer directory zip test"
echo "$(ls viewerDirectoryZipTest)"
```
In the created `viewerDirectoryZipTest` you should see `ixbrlviewer.zip` and when unzipped you should see `ixbrlviewer.htm` and `filing-documents.zip`. 

5. Test load from a zip and output to a file as a zip
```bash
mkdir viewerFileZipTest
python arelleCmdLine.py --plugins=<PATH_TO_VIEWER>/iXBRLViewerPlugin -f <PATH_TO_DOWNLOADS>/filing_documents.zip  --save-viewer <PATH_TO_OUTPUT>/viewerFileZipTest/ixbrlviewer.zip --viewer-url <PATH_TO_VIEWER>/XBRLViewerPlugin/viewer/dist/ixbrlviewer.js --zip-viewer-output
echo ""
echo "Viewer file zip test"
echo "$(ls viewerFileZipTest)"
```

In the created `viewerFileZipTest` you should see `ixbrlviewer.zip` and when unzipped you should see `ixbrlviewer.htm` and `filing-documents.zip`. 

#### Test with inline doc entrypoint commandline
For all of the following you will expect:
1. Only the viewer file(s) to be outputted
2. Not have the download filing documents button available under the gear icon
3. Not see an entry for `filingDocuments` In the ixbrl-viewer.json

##### Testing Steps
1. Unzip provided filing documents
2. Test load from an inline doc and output to a directory
```
mkdir viewerDirectoryTestInlineDoc
python arelleCmdLine.py --plugins=<PATH_TO_VIEWER>/iXBRLViewerPlugin -f <PATH_TO_DOWNLOADS>/ixbrlviewertest-20230502.htm  --save-viewer <PATH_TO_OUTPUT>/viewerDirectoryTestInlineDoc --viewer-url <PATH_TO_VIEWER>/XBRLViewerPlugin/viewer/dist/ixbrlviewer.js
echo ""
echo "Viewer directory test"
echo "$(ls viewerDirectoryTestInlineDoc)"
```

In the created `viewerDirectoryTestInlineDoc` you should see `ixbrlviewer.htm`. 

3. Test load from an inline doc and output to a file
```bash
mkdir viewerFileTestInlineDoc
python arelleCmdLine.py --plugins=<PATH_TO_VIEWER>/iXBRLViewerPlugin -f <PATH_TO_DOWNLOADS>/ixbrlviewertest-20230502.htm  --save-viewer <PATH_TO_OUTPUT>/viewerFileTestInlineDoc/ixbrlviewer.html --viewer-url <PATH_TO_VIEWER>/iXBRLViewerPlugin/viewer/dist/ixbrlviewer.js
echo ""
echo "Viewer file test"
echo "$(ls viewerFileTestInlineDoc)"
```

In the created `viewerFileTestInlineDoc` you should see `ixbrlviewer.htm`. 

4. Test load from an inline doc and output to a directory as a zip
```bash
mkdir viewerDirectoryZipTestInlineDoc
python arelleCmdLine.py --plugins=<PATH_TO_VIEWER>/iXBRLViewerPlugin -f <PATH_TO_DOWNLOADS>/ixbrlviewertest-20230502.htm  --save-viewer <PATH_TO_OUTPUT>/viewerDirectoryZipTestInlineDoc --viewer-url <PATH_TO_VIEWER>/iXBRLViewerPlugin/viewer/dist/ixbrlviewer.js --zip-viewer-output
echo ""
echo "Viewer directory zip test"
echo "$(ls viewerDirectoryZipTestInlineDoc)"
```

In the created `viewerDirectoryZipTestInlineDoc` you should see `ixbrlviewer.zip` and when unzipped you should see `ixbrlviewer.htm`. 

5. Test load from an inline doc and output to a file as a zip
```bash
mkdir viewerFileZipTestInlineDoc
python arelleCmdLine.py --plugins=<PATH_TO_VIEWER>/iXBRLViewerPlugin -f <PATH_TO_DOWNLOADS>/ixbrlviewertest-20230502.htm  --save-viewer <PATH_TO_OUTPUT>/viewerFileZipTestInlineDoc/ixbrlviewer.zip --viewer-url <PATH_TO_VIEWER>/XBRLViewerPlugin/viewer/dist/ixbrlviewer.js --zip-viewer-output
echo ""
echo "Viewer file zip test"
echo "$(ls viewerFileZipTestInlineDoc)"
```

In the created `viewerFilTestInlineDoc` you should see `ixbrlviewer.zip` and when unzipped you should see `ixbrlviewer.htm`. 

### Test the arelle gui
1. Pull this branch
2. Build the viewer `npm install && nmp run prod`
3. Do all of the following from your local arelle
4. Start arelle gui `python arelleGUI.pyw`
5. In the arelle menu select help
6. Select manage plugins
7. Select `<PATH_TO_VIEWER>/iXBRLViewerPlugin/__init__.py`
8. Enable
9. Close
10. Reload

#### Testing
For all of the tests you will run `python arelleGUI.pyw` to start arelle

##### Test load from zip entrypoint don't output as zip 
1. Click the file icon
2. Select the <PATH_TO_DOWNLOADS>/filingdocuments.zip
3. Select the htm file in the zip
4. After a few moments verify that the viewer opened in the browser (currently broken on master)
5. Under the gear icon verify that download filing documents is available
6. Verify that the downloaded docs match the original docs in both content and name
7. Navigate back to arelle
8. Select tools
9. Select save viewer
10. Enter `<PATH_TO_VIEWER>/XBRLViewerPlugin/viewer/dist/ixbrlviewer.js` as your viewer url
11. Enter `<PATH_TO_OUTPUT>/ixbrlviewer.htm` as your iXBRL File
12. Unselect `Zip Viewer Output`
13. Save
14. Verify that `ixbrlviewer.htm` and `filingdocuments.zip` are saved in `<PATH_TO_OUTPUT>`
15. Open `ixbrlviewer.htm` verify that download filing documents is available under the gear
16. Verify that the downloaded filing documents match the original docs in both name and contents

##### Test load from zip entrypoint output as zip 
1. Click the file icon
2. Select the <PATH_TO_DOWNLOADS>/filingdocuments.zip
3. Select the htm file in the zip
4. After a few moments verify that the viewer opened in the browser (currently broken on master)
5. Under the gear icon verify that download filing documents is available
6. Verify that the downloaded docs match the original docs in both content and name
7. Navigate back to arelle
8. Select tools
9. Select save viewer
10. Enter `<PATH_TO_VIEWER>/XBRLViewerPlugin/viewer/dist/ixbrlviewer.js` as your viewer url
11. Enter `<PATH_TO_OUTPUT>/ixbrlviewer.zip` as your iXBRL File
12. Select `Zip Viewer Output`
13. Save
14. Verify that `ixbrlviewer.htm` and `filingdocuments.zip` are saved in `<PATH_TO_OUTPUT>/ixbrlviewer.zip`
15. Open `ixbrlviewer.htm` verify that download filing documents is available under the gear
16. Verify that the downloaded filing documents match the original docs in both name and contents

##### Test load from ixbrl doc entrypoint don't output as zip 
1. Click the file icon
2. Select the <PATH_TO_DOWNLOADS>/ixbrlviewertest-20230502.htm
3. After a few moments verify that the viewer opened in the browser (currently broken on master)
4. Under the gear icon verify that download filing documents is not available
5. Navigate back to arelle
6. Select tools
7. Select save viewer
8. Enter `<PATH_TO_VIEWER>/XBRLViewerPlugin/viewer/dist/ixbrlviewer.js` as your viewer url
9. Enter `<PATH_TO_OUTPUT>/ixbrlviewer.htm` as your iXBRL File
10. Unselect `Zip Viewer Output`
11. Save
12. Verify that `ixbrlviewer.htm` is saved in `<PATH_TO_OUTPUT>`
13. Open `ixbrlviewer.htm` verify that download filing documents is not available under the gear

##### Test load from zip entrypoint output as zip 
1. Click the file icon
2. Select the <PATH_TO_DOWNLOADS>/ixbrlviewertest-20230502.htm
3. After a few moments verify that the viewer opened in the browser (currently broken on master)
4. Under the gear icon verify that download filing documents is not available
5. Navigate back to arelle
6. Select tools
7. Select save viewer
8. Enter `<PATH_TO_VIEWER>/XBRLViewerPlugin/viewer/dist/ixbrlviewer.js` as your viewer url
9. Enter `<PATH_TO_OUTPUT>/ixbrlviewer.zip` as your iXBRL File
10. Select `Zip Viewer Output`
11. Save
12. Verify that `ixbrlviewer.htm` is saved in `<PATH_TO_OUTPUT>`
13. Open `ixbrlviewer.htm` verify that download filing documents is not available under the gear

### Test IXDS
Use [ixds.zip](https://github.com/Workiva/ixbrl-viewer/files/11378223/ixds.zip)

#### Testing
1. Test load ixds from zip and save to zip
```bash
mkdir viewerIXDSZipTestZipToFile
python arelleCmdLine.py --plugins '<PATH_TO_VIEWER>/XBRLViewerPlugin|inlineXbrlDocumentSet' -f '[{"ixds":[{"file":"<PATH_TO_DOWNLOADS>/ixds.zip/ixbrlviewertest-20230502.htm"},{"file":"<PATH_TO_DOWNLOADS>/ixds.zip/ixbrlviewertest-20230502_d2.htm"}]}]'  --save-viewer <PATH_TO_OUTPUT>/viewerIXDSZipTestZipToFile/ixbrlviewer.zip --viewer-url <PATH_TO_VIEWER>/iXBRLViewerPlugin/viewer/dist/ixbrlviewer.js --zip-viewer-output
echo ""
echo "Viewer file zip test"
echo "$(ls viewerIXDSZipTestZipToFile)"
```
Expect `viewerIXDSZipTestZipToFile` to contain `ixbrlviewer.zip` that when unzipped contains 2 `.htm` files, `ixds.zip`. Launch a python simple server `python -m http.server` from the unzipped directory in `viewerIXDSZipTestZipToFile` and in the browser navigate to localhost 8080 and click on the main `htm`. Verify that download filing docs is available under the gear icon and returns a zip with the same name and contents as the original filing docs.

2. Test load ixds from zip and save zip to dir
mkdir viewerIXDSZipTestZipToDir
python arelleCmdLine.py --plugins '<PATH_TO_VIEWER>/iXBRLViewerPlugin|inlineXbrlDocumentSet' -f '[{"ixds":[{"file":"<PATH_TO_DOWNLOADS>/ixds.zip/ixbrlviewertest-20230502.htm"},{"file":"<PATH_TO_DOWNLOADS>/ixds.zip/ixbrlviewertest-20230502_d2.htm"}]}]'  --save-viewer <PATH_TO_OUTPUT>/viewerIXDSZipTestZipToDir --viewer-url <PATH_TO_VIEWER>/iXBRLViewerPlugin/viewer/dist/ixbrlviewer.js --zip-viewer-output
echo ""
echo "Viewer file zip test"
echo "$(ls viewerIXDSZipTestZipToDir)"
```
Expect `viewerIXDSZipTestZipToDir` to contain `ixbrlviewer.zip` that when unzipped contains 2 `.htm` files, `ixds.zip`. Launch a python simple server `python -m http.server` from the unzipped dir in `viewerIXDSZipTestZipToDir` and in the browser navigate to localhost 8080 and click on the main `htm`. Verify that download filing docs is available under the gear icon and returns a zip with the same name and contents as the original filing docs.

3. Test load ixds from zip and save output to dir
```
mkdir viewerIXDSZipTestSaveToDir
python arelleCmdLine.py --plugins '<PATH_TO_VIEWER>/iXBRLViewerPlugin|inlineXbrlDocumentSet' -f '[{"ixds":[{"file":"<PATH_TO_DOWNLOADS>/ixds.zip/ixbrlviewertest-20230502.htm"},{"file":"<PATH_TO_DOWNLOADS>/ixds.zip/ixbrlviewertest-20230502_d2.htm"}]}]'  --save-viewer <PATH_TO_OUTPUT>/viewerIXDSZipTestSaveToDir --viewer-url <PATH_TO_VIEWER>/iXBRLViewerPlugin/viewer/dist/ixbrlviewer.js
echo ""
echo "Viewer file zip test"
echo "$(ls viewerIXDSZipTestSaveToDir)"
```

Expect `viewerIXDSZipTestSaveToDir` to contain 2 `.htm` files, `ixds.zip`. Launch a python simple server `python -m http.server` from `viewerIXDSZipTestSaveToDir` and in the browser navigate to localhost 8080 and click on the main `htm`. Verify that download filing docs is available under the gear icon and returns a zip with the same name and contents as the original filing docs.

4. Test load ixds and zip to file
```
mkdir viewerIXDSTestZipToFile
python arelleCmdLine.py --plugins '<PATH_TO_VIEWER>/iXBRLViewerPlugin|inlineXbrlDocumentSet' -f '[{"ixds":[{"file":"<PATH_TO_DOWNLOADS>/ixds/ixbrlviewertest-20230502.htm"},{"file":"<PATH_TO_DOWNLOADS>/ixds/ixbrlviewertest-20230502_d2.htm"}]}]'  --save-viewer <PATH_TO_OUTPUT>/viewerIXDSTestZipToFile/ixbrlviewer.zip --viewer-url <PATH_TO_VIEWER>/iXBRLViewerPlugin/viewer/dist/ixbrlviewer.js --zip-viewer-output
echo ""
echo "Viewer file zip test"
echo "$(ls viewerIXDSTestZipToFile)"
```

Expect `viewerIXDSTestZipToFile` to contain `ixbrlviewer.zip` that when unzipped contains 2 `.htm` files. Launch a python simple server `python -m http.server` from  the unzipped dir in `viewerIXDSTestZipToFile` and in the browser navigate to localhost 8080 and click on the main `htm`. Verify that download filing docs is not available under the gear icon.

5. Test load ixds and zip to dir
```bash
mkdir viewerIXDSTestZipToDir
python arelleCmdLine.py --plugins '<PATH_TO_VIEWER>/iXBRLViewerPlugin|inlineXbrlDocumentSet' -f '[{"ixds":[{"file":"<PATH_TO_DOWNLOADS>/ixds/ixbrlviewertest-20230502.htm"},{"file":"<PATH_TO_DOWNLOADS>/xds/ixbrlviewertest-20230502_d2.htm"}]}]'  --save-viewer <PATH_TO_OUTPUT>/viewerIXDSTestZipToDir --viewer-url <PATH_TO_VIEWER>/iXBRLViewerPlugin/viewer/dist/ixbrlviewer.js --zip-viewer-output
echo ""
echo "Viewer file zip test"
echo "$(ls viewerIXDSTestZipToDir)"
```

Expect `viewerIXDSTestZipToDir` to contain `ixbrlviewer.zip` that when unzipped contains 2 `.htm` files. Launch a python simple server `python -m http.server` from the unzipped dir in `viewerIXDSTestZipToDir` and in the browser navigate to localhost 8080 and click on the main `htm`. Verify that download filing docs is not available under the gear icon.

6. Test load ixds and save to dir
```bash
mkdir viewerIXDSTestSaveToDir
python arelleCmdLine.py --plugins '<PATH_TO_VIEWER>/iXBRLViewerPlugin|inlineXbrlDocumentSet' -f '[{"ixds":[{"file":"<PATH_TO_DOWNLOADS>/ixds/ixbrlviewertest-20230502.htm"},{"file":"<PATH_TO_DOWNLOADS>/ixds/ixbrlviewertest-20230502_d2.htm"}]}]'  --save-viewer <PATH_TO_OUTPUT>/viewerIXDSTestSaveToDir --viewer-url <PATH_TO_VIEWER>/iXBRLViewerPlugin/viewer/dist/ixbrlviewer.js
echo ""
echo "Viewer file zip test"
echo "$(ls viewerIXDSTestSaveToDir)"
```

Expect `viewerIXDSTestSaveToDir` to contain 2 `.htm` files. Launch a python simple server `python -m http.server` from `viewerIXDSZipTestZipToFile` and in the browser navigate to localhost 8080 and click on the main `htm`. Verify that download filing docs is  not available under the gear icon.

**review**:
@Workiva/xt
@paulwarren-wk

### CODE REVIEW

**Comment by aviary2-wf** [comment]
## Security Insights
The items listed below may not capture all security relevant changes. Before providing a security review, be sure to review the entire PR for security impact.
<details>
<summary>(1) Security relevant changes were detected</summary>
<li>Watched keyword <code>etree</code> in <code>iXBRLViewerPlugin/iXBRLViewer.py</code> line(s) ['18'] removed</li>
</details>

## Action Items
* Request infosec review via Slack: #support-infosec
* Verify [`aviary.yaml`](https://wiki.atl.workiva.net/pages/viewpage.action?pageId=42762752#SecurityInsightFramework&Tooling-ConfiguringAviary) coverage of security relevant code
---
_Questions or Comments? Reach out on Slack: #support-infosec._

File: iXBRLViewerPlugin/__init__.py
Comment (derekgengenbacher-wf) [inline iXBRLViewerPlugin/__init__.py:169]: Why did you do this? Why not pull this function out into its own function with documentation?

File: iXBRLViewerPlugin/__init__.py
Comment (derekgengenbacher-wf) [inline iXBRLViewerPlugin/__init__.py:190]: It looks like this never gets cleaned up. Perhaps instantiate this with a `with` statement?

File: iXBRLViewerPlugin/__init__.py
Comment (aaroncameron-wk) [inline iXBRLViewerPlugin/__init__.py:58]: This value is used in at least 3 places elsewhere. Could we set a constant and reuse it?

File: iXBRLViewerPlugin/__init__.py
Comment (sagesmith-wf) [reply iXBRLViewerPlugin/__init__.py:169]: I can pull the class out if you want, but I just moved some code. `getLocalFile` is an override to `LocalViewer.getLocalFile`

File: iXBRLViewerPlugin/__init__.py
Comment (sagesmith-wf) [reply iXBRLViewerPlugin/__init__.py:58]: Sure

File: iXBRLViewerPlugin/__init__.py
Comment (austinmatherne-wk) [inline iXBRLViewerPlugin/__init__.py:47]: nit: move this up with the rest of the imports

File: iXBRLViewerPlugin/__init__.py
Comment (austinmatherne-wk) [inline iXBRLViewerPlugin/__init__.py]: I'm not sure what you're trying to say at the end of this comment. Did Thor hijack your keyboard? 😄

File: iXBRLViewerPlugin/__init__.py
Comment (austinmatherne-wk) [inline iXBRLViewerPlugin/__init__.py]: Do these paths work on Windows?

File: iXBRLViewerPlugin/__init__.py
Comment (austinmatherne-wk) [reply iXBRLViewerPlugin/__init__.py]: 👨‍🍳

File: iXBRLViewerPlugin/__init__.py
Comment (sagesmith-wf) [reply iXBRLViewerPlugin/__init__.py:47]: Pre-existing: But cando

File: iXBRLViewerPlugin/iXBRLViewer.py
Comment (austinmatherne-wk) [inline iXBRLViewerPlugin/iXBRLViewer.py:400]: You forgot to remove this

**Review by austinmatherne-wk — APPROVED** [review]

**Comment by austinmatherne-wk** [comment]
QA +10
@Workiva/release-management-p

**Comment by soufianeelaissaoui-wk** [comment]
Looks good to me. The etree from lxml looks safe. Security +1

**Comment by colefeisthamel-wf** [comment]
RM +1 approvals added by team after merge

**Pull request merged**
